socketshark/receiver.py

```diff
@@ -79,31 +78,37 @@ async def ping_handler(self):
             if wait:
                 wait.cancel()
             if not self._stop:
-                self.shark.log.exception('unhandled exception in ping handler')
+                self.shark.log.debug('redis ping handler cancelled')
         except Exception:
             self.shark.log.exception('unhandled exception in ping handler')
+            self._stop = True
         finally:
-            await self.stop()
+            if self._stop:
+                await self.stop()
 
     async def reader(self, once=False):
         self._stop = False
         try:
             ping_handler = asyncio.ensure_future(self.ping_handler())
-            result = await self._reader(once=once)
+            tasks = [
+                self._reader(connection, once=once)
+                for connection in self.redis_connections
+            ]
+            result = await asyncio.gather(*tasks)
             ping_handler.cancel()
             return result
         except Exception:
             self.shark.log.exception('unhandled exception in receiver')
 
-    async def _reader(self, once=False):
-        prefix_length = len(self.redis_channel_prefix)
-        if once and not self.redis_receiver._queue.qsize():
+    async def _reader(self, connection, once=False):
```

Comment:
Nit: Maybe let's call it `_reader_for_connection`?

Reply:
Done in a415332.